src/irc/core/sasl.c

```diff
@@ -105,49 +113,149 @@ static void sasl_success(IRC_SERVER_REC *server, const char *data, const char *f
+	 * is expected.
+	 */
+	if (strlen(enc_req) == AUTHENTICATE_CHUNK_SIZE) {
+		g_hash_table_insert(sasl_buffers, server->tag, g_strdup(enc_req));
+		g_free_not_null(buffer);
+		return NULL;
+	}
 
-	/* Stop the timer */
-	if (server->sasl_timeout != 0) {
-		g_source_remove(server->sasl_timeout);
-		server->sasl_timeout = 0;
+	if (g_strcmp0("+", enc_req) == 0) {
+		decoded = g_bytes_new("", 0);
+	} else {
+		gsize dec_len;
+		guchar *tmp;
+
+		tmp = g_base64_decode(enc_req, &dec_len);
+		decoded = g_bytes_new_take(tmp, dec_len);
+	}
+
+	g_free_not_null(buffer);
+	return decoded;
+}
+
+/*
+ * Splits the response into appropriately sized chunks for the AUTHENTICATION
+ * command to be sent to the IRC server. If |response| is NULL, then the empty
+ * response is sent to the server.
+ */
+void sasl_send_response(IRC_SERVER_REC *server, GBytes *response)
+{
+	char *enc;
+	const guchar *p;
+	gsize len;
+	size_t offset, enc_len, chunk_len;
+
+	if (response == NULL) {
+		irc_send_cmdv(server, "AUTHENTICATE +");
+		return;
+	}
+
+	p = g_bytes_get_data(response, &len);
+	enc = g_base64_encode(p, len);
+	enc_len = strlen(enc);
+
+	for (offset = 0; offset < enc_len; offset += AUTHENTICATE_CHUNK_SIZE) {
+		chunk_len = enc_len - offset;
+		if (chunk_len > AUTHENTICATE_CHUNK_SIZE)
+			chunk_len = AUTHENTICATE_CHUNK_SIZE;
+
+		irc_send_cmdv(server, "AUTHENTICATE %.*s", (int) chunk_len, enc + offset);
+	}
+
+	if (offset == enc_len) {
+		irc_send_cmdv(server, "AUTHENTICATE +");
 	}
+	g_free(enc);
+}
+
+/*
+ * Called when the incoming SASL request is completely received.
+ */
+static void sasl_step_complete(IRC_SERVER_REC *server, GBytes *data)
+{
+	IRC_SERVER_CONNECT_REC *conn;
+	GString *resp;
+	gsize resp_len;
+	GBytes *resp_bytes;
+
+	conn = server->connrec;
 
 	switch (conn->sasl_mechanism) {
 		case SASL_MECHANISM_PLAIN:
 			/* At this point we assume that conn->sasl_{username, password} are non-NULL.
 			 * The PLAIN mechanism expects a NULL-separated string composed by the authorization identity, the
 			 * authentication identity and the password.
 			 * The authorization identity field is explicitly set to the user provided username.
-			 * The whole request is then encoded in base64. */
+			 */
 
-			req = g_string_new(NULL);
+			resp = g_string_new(NULL);
 
-			g_string_append(req, conn->sasl_username);
-			g_string_append_c(req, '\0');
-			g_string_append(req, conn->sasl_username);
-			g_string_append_c(req, '\0');
-			g_string_append(req, conn->sasl_password);
+			g_string_append(resp, conn->sasl_username);
+			g_string_append_c(resp, '\0');
+			g_string_append(resp, conn->sasl_username);
+			g_string_append_c(resp, '\0');
+			g_string_append(resp, conn->sasl_password);
 
-			enc_req = g_base64_encode((const guchar *)req->str, req->len);
+			/* This could be g_string_free_to_bytes */
+			resp_len = resp->len;
+			resp_bytes = g_bytes_new_take(g_string_free(resp, FALSE), resp_len);
```

Comment:
Added in glib 2.32 (current requirement is 2.16, shouldn't be bumped past 2.28 because rhel 6)